[Win] Link clang_rt.builtins in preparation for int128_t

rkirsling opened the pull request:
#### 25ceab605c309e2349da68beff0c75a3485ad48c
<pre>
[Win] Link clang_rt.builtins in preparation for int128_t
<a href="https://bugs.webkit.org/show_bug.cgi?id=274512">https://bugs.webkit.org/show_bug.cgi?id=274512</a>

Reviewed by Yusuke Suzuki.

* Source/cmake/OptionsMSVC.cmake:

Canonical link: <a href="https://commits.webkit.org/279260@main">https://commits.webkit.org/279260@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b24c0282093fa2a5a2c9328fdd7852455d421219

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/52936 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/32273 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/5423 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/56215 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/3659 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/39036 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/3384 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/42944 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/2356 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/55034 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/29981 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/45695 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/24062 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/27049 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/1818 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/46290 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/48904 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/3171 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/57808 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/52449 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/28077 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/3137 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/50339 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/29297 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/45912 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/49634 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/11557 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/30216 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/64753 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/29051 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/12292 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

fujii commented on Source/cmake/OptionsMSVC.cmake:
```diff
@@ -192,9 +192,12 @@ string(REPLACE "INCREMENTAL:YES" "INCREMENTAL:NO" replace_CMAKE_EXE_LINKER_FLAGS
-    # FIXME: Building with clang-cl seemed to fail with 128 bit int support
-    set(HAVE_INT128_T OFF)
-    list(REMOVE_ITEM _WEBKIT_CONFIG_FILE_VARIABLES HAVE_INT128_T)
+    get_filename_component(CMAKE_CXX_COMPILER_DIR ${CMAKE_CXX_COMPILER} DIRECTORY)
```
I think variable names starting with `CMAKE_` is reserved for CMake. We should avoid such names.

fujii commented on Source/cmake/OptionsMSVC.cmake:
```diff
@@ -192,9 +192,12 @@ string(REPLACE "INCREMENTAL:YES" "INCREMENTAL:NO" replace_CMAKE_EXE_LINKER_FLAGS
-    list(REMOVE_ITEM _WEBKIT_CONFIG_FILE_VARIABLES HAVE_INT128_T)
+    get_filename_component(CMAKE_CXX_COMPILER_DIR ${CMAKE_CXX_COMPILER} DIRECTORY)
+    string(REGEX MATCH "^[0-9]+" CMAKE_CXX_COMPILER_MAJOR_VERSION ${CMAKE_CXX_COMPILER_VERSION})
+    string(JOIN "/" CLANG_BUILTINS_LIBRARY
```
I think you should use `cmake_path(APPEND)` and `find_library` with `REQUIRED` flag.
https://cmake.org/cmake/help/latest/command/cmake_path.html#append
https://cmake.org/cmake/help/latest/command/find_library.html

fujii commented on Source/cmake/OptionsMSVC.cmake:
```diff
@@ -192,9 +192,12 @@ string(REPLACE "INCREMENTAL:YES" "INCREMENTAL:NO" replace_CMAKE_EXE_LINKER_FLAGS
+    string(JOIN "/" CLANG_BUILTINS_LIBRARY
+        ${CMAKE_CXX_COMPILER_DIR} "../lib/clang" ${CMAKE_CXX_COMPILER_MAJOR_VERSION} "lib/windows/clang_rt.builtins-x86_64.lib")
+    string(APPEND CMAKE_SHARED_LINKER_FLAGS " \"${CLANG_BUILTINS_LIBRARY}\"")
+    string(APPEND CMAKE_EXE_LINKER_FLAGS " \"${CLANG_BUILTINS_LIBRARY}\"")
```
Can we use `add_library` here?
https://cmake.org/cmake/help/latest/command/add_library.html

rkirsling replied:
I'm not sure how that would work; I'm just doing the same thing we do for ASan:
https://github.com/WebKit/WebKit/blob/main/Source/cmake/WebKitCompilerFlags.cmake#L264

Constellation reviewed (approved):
r=me

rkirsling committed:
[Win] Link clang_rt.builtins in preparation for int128_t
https://bugs.webkit.org/show_bug.cgi?id=274512

Reviewed by Yusuke Suzuki.

* Source/cmake/OptionsMSVC.cmake:

Canonical link: https://commits.webkit.org/279260@main
